addons/account/account_payment.py

```diff
@@ -0,0 +1,378 @@
+                self.destination_account_id = self.partner_id.property_account_receivable.id
+            else:
+                self.destination_account_id = self.partner_id.property_account_payable.id
+
+    @api.one
+    @api.depends('invoice_id', 'amount')
+    def _compute_payment_difference(self):
+        self.payment_difference = self.invoice_id and self.invoice_id.residual - self.amount or 0
+
+    name = fields.Char(readonly=True, copy=False)
+    state = fields.Selection([('draft','Draft'), ('confirmed','Confirmed')], readonly=True, default='draft')
+
+    payment_type = fields.Selection([('outbound', 'Send Money'), ('inbound', 'Receive Money'), ('transfer', 'Transfer Money')], default='outbound', required=True)
+    payment_method = fields.Many2one('account.payment.method', string='Payment Method', required=True)
+    # default True to prevent a visual glitch
+    hide_payment_method = fields.Boolean(default=True, copy=False)
+    payment_state = fields.Selection([('todo', 'To Process'), ('done', 'Processed'), ('failed', 'Failed')], required=True, default="todo")
+
+    partner_type = fields.Selection([('customer', 'Customer'), ('supplier', 'Supplier')], default='supplier')
+    partner_id = fields.Many2one('res.partner', string='Partner')
+
+    amount = fields.Float(string='Amount', required=True, digits=0)
+    date = fields.Date(string='Date', default=fields.Date.context_today, required=True, copy=False)
+    reference = fields.Char('Ref', help="Transaction reference number.")
+    journal_id = fields.Many2one('account.journal', string='Cash / Bank Journal', required=True, domain=[('type', 'in', ('bank', 'cash'))])
+    # Money flows from the journal_id's default_debit_account_id or default_credit_account_id to the destination_account_id
+    destination_account_id = fields.Many2one('account.account', compute='_compute_destination_account_id', required=True)
+    # For money transfer, money goes from journal_id to a transfer account, then from the transfer account to destination_journal_id
+    destination_journal_id = fields.Many2one('account.journal', string='Transfer To', domain=[('type', 'in', ('bank', 'cash'))])
+    company_id = fields.Many2one('res.company', related='journal_id.company_id', string='Company', readonly=True)
+
+    invoice_id = fields.Many2one('account.invoice', string="Invoice", domain=[('state', '=', 'open')])
+    payment_difference = fields.Float(compute='_compute_payment_difference')
+    payment_difference_handling = fields.Selection([('open', 'Keep Open'), ('reconcile', 'Reconcile Payment Balance')], default='open', string="Payment Difference")
+    writeoff_account = fields.Many2one('account.account', string="Counterpart Account", domain=[('deprecated', '=', False)])
+
+    move_lines = fields.One2many('account.move.line', 'payment_id', copy=False, readonly=True, ondelete='restrict')
+
+    @api.one
+    @api.constrains('amount')
+    def _check_amount(self):
+        if self.amount == 0.0:
```

Comment:
self.amont > 0:
  raise ValidationError('The payment amount must be strictly positive.')